File: llm/steps/memory.py
"""
Step 4: MEMORY - Archivist.
Updates the rolling summary after Mouth responds.
"""
import logging
import time
from typing import Tuple, Optional
from llm.schemas import PipelineInput, BrainOutput, MouthOutput, MemoryOutput
from llm.api_helpers import make_api_call
from llm.prompts import MEMORY_SYSTEM_PROMPT, MEMORY_USER_TEMPLATE

logger = logging.getLogger(__name__)


# JSON Schema for Memory output (inline)
MEMORY_SCHEMA = {
    "name": "memory_output",
    "strict": False,
    "schema": {
        "type": "object",
        "properties": {
            "updated_rolling_summary": {
                "type": "string",
                "description": "Updated summary (80-200 words)"
            },
            "needs_recursive_summary": {
                "type": "boolean"
            }
        },
        "required": ["updated_rolling_summary", "needs_recursive_summary"],
        "additionalProperties": False
    }
}


async def run_memory(
    context: PipelineInput,
    user_message: str,
    mouth_output: Optional[MouthOutput],
    brain_output: BrainOutput,
    tracer: Optional[object] = None
) -> Optional[str]:
    """
    Run the Memory step (Async).
    Returns the new summary string so the worker can save it.
    Runs AFTER Mouth output is available.
    """
    try:
        data, usage_dict, latency = await _run_memory_llm(
            context, user_message, mouth_output, brain_output
        )

        output = MemoryOutput(
            updated_rolling_summary=data.get("updated_rolling_summary", ""),
            needs_recursive_summary=data.get("needs_recursive_summary", False),
        )

        if tracer:
            tracer.log_step(
                step_name="Memory",
                input_data={"user_message": user_message[:50]},
                output_data=data,
                latency_ms=latency,
                model="llama3-70b-8192",
                token_usage=usage_dict
            )

        logger.info(f"Memory: {len(output.updated_rolling_summary)} chars")
        return output.updated_rolling_summary
        
    except Exception as e:
        logger.error(f"Memory failed: {e}", exc_info=True)
        return context.rolling_summary or "No summary available"


async def _run_memory_llm(
    context: PipelineInput,
    user_message: str,
    mouth_output: Optional[MouthOutput],
    brain_output: BrainOutput
) -> Tuple[dict, dict, int]:
    """Core LLM Logic."""
    bot_message = mouth_output.message_text if mouth_output else "(No response sent)"
    action_taken = f"Action: {brain_output.action.value}, Stage: {brain_output.new_stage.value}"
    
    user_prompt = MEMORY_USER_TEMPLATE.format(
        rolling_summary=context.rolling_summary or "No prior summary",
        user_message=user_message,
        bot_message=bot_message,
        action_taken=action_taken,
    )
    
    start_time = time.time()

    data, usage = await make_api_call(
        messages=[
            {"role": "system", "content": MEMORY_SYSTEM_PROMPT},
            {"role": "user", "content": user_prompt},
        ],
        response_format={"type": "json_schema", "json_schema": MEMORY_SCHEMA},
        max_tokens=2000,
        step_name="Memory",
        strict=False
    )
    
    latency = int((time.time() - start_time) * 1000)
    return data, usage, latency

File: llm-go/steps/memory/memory.go
package memory

import (
	"fmt"
	"time"
	"whatsapp-funnel/llm-go/apihelpers"
	"whatsapp-funnel/llm-go/config"
	"whatsapp-funnel/llm-go/prompts"
	"whatsapp-funnel/llm-go/schemas"
)

var memorySchema = map[string]interface{}{
	"name":   "memory_output",
	"strict": false,
	"schema": map[string]interface{}{
		"type": "object",
		"properties": map[string]interface{}{
			"updated_rolling_summary": map[string]string{"type": "string"},
			"needs_recursive_summary": map[string]string{"type": "boolean"},
		},
	},
}

func RunMemory(cfg config.LLMConfig, context schemas.PipelineInput, userMessage string, generateOutput schemas.GenerateOutput) string {
	output, _, err := runMemoryLLM(cfg, context, userMessage, generateOutput)
	if err != nil {
		if context.RollingSummary != "" {
			return context.RollingSummary
		}
		return "No summary available"
	}
	return output.UpdatedRollingSummary
}

func runMemoryLLM(cfg config.LLMConfig, context schemas.PipelineInput, userMessage string, generateOutput schemas.GenerateOutput) (schemas.MemoryOutput, int, error) {
	botMessage := generateOutput.MessageText
	if botMessage == "" {
		botMessage = "(No response sent)"
	}
	actionTaken := fmt.Sprintf("Action: %s, Stage: %s", generateOutput.Action, generateOutput.NewStage)
	summary := context.RollingSummary
	if summary == "" {
		summary = "No prior summary"
	}
	prompt := fmt.Sprintf(prompts.MemoryUserTemplate, summary, userMessage, botMessage, actionTaken)
	start := time.Now()
	data, err := apihelpers.MakeAPICall(cfg,
		[]apihelpers.Message{{Role: "system", Content: prompts.MemorySystemPrompt}, {Role: "user", Content: prompt}},
		map[string]interface{}{"type": "json_schema", "json_schema": memorySchema},
		0.7, intPtr(2000), "Memory", false,
	)
	if err != nil {
		return schemas.MemoryOutput{}, int(time.Since(start).Milliseconds()), err
	}
	out := schemas.MemoryOutput{}
	if v, ok := data["updated_rolling_summary"].(string); ok {
		out.UpdatedRollingSummary = v
	}
	if v, ok := data["needs_recursive_summary"].(bool); ok {
		out.NeedsRecursiveSummary = v
	}
	return out, int(time.Since(start).Milliseconds()), nil
}

func intPtr(v int) *int { return &v }
